### ISSUE TITLE
[BUG] Unable to select NIC to create network config when cluster contains witness node

### ISSUE THREAD

**Issue by albinsun** [harvester/harvester#5325]
We found that NIC does not shown in the network config creating page when cluster contains a witness node.
It works in single and two default nodes cluster but not when witnes exists (We hit in both DW and DDW)

## Environment
* Harvester
  * Version: `v1.3.0-rc4`
  * Profile: QEMU/KVM
  * ui-source: `Auto`

## Reproduce Steps 
1. Create harvester with single default node
   * <details>
     <summary> Can select NIC when creating network config
     </summary>

     ![image](https://github.com/harvester/harvester/assets/2773781/e1126448-aa2f-4c29-9288-23b993b6ebb9)
     </details>
   
1. Join a witness node
   * <details open>
     <summary>  Can select NIC when creating network config  :red_circle: ${\color{red} \text{No NICs shown}}$
     </summary>

     ![image](https://github.com/harvester/harvester/assets/2773781/e5d3045e-9bc2-4b0f-a37e-cc5d1938374b)
     </details>

1. Delete witness node and join a default node
   * <details>
     <summary> Delete witness node
     </summary>

     ![image](https://github.com/harvester/harvester/assets/2773781/1ef8548d-d2f8-4340-be55-5056a709571b)
     </details>

   * <details>
     <summary> Can select NIC when creating network config
     </summary>

     ![image](https://github.com/harvester/harvester/assets/2773781/1ac88d3d-79ba-468e-99b7-16b8d07bc3e9)
     </details>

## Note
1. Witness node do have the interface
   ![image](https://github.com/harvester/harvester/assets/2773781/8ebafeb2-f805-47af-b075-b09b9a45b3bb)

## Supportbundle
[supportbundle_nicWhenWitness_2024-03-08T08-53-12Z.zip](https://github.com/harvester/harvester/files/14535919/supportbundle_nicWhenWitness_2024-03-08T08-53-12Z.zip)

**Comment by w13915984028** [comment]
@albinsun Please try to set the `Node Selector` first, let it select the none-witness node, then set 'UpLink`.

**Comment by albinsun** [comment]
@w13915984028, It works! Thank you!

Decreasing the severity to 3.

![image](https://github.com/harvester/harvester/assets/2773781/901a9c44-865a-42ce-8712-1bfbfd7807f0)
![image](https://github.com/harvester/harvester/assets/2773781/bf580ecf-31c9-4cd9-8428-daf655efa96b)

**Comment by bk201** [comment]
Move to v1.4.0, we can add a document for v1.3.0: https://github.com/harvester/harvester/issues/5333
@torchiaf please help survey the reason the common uplink can't be selected from GUI's perspective.

**Comment by a110605** [comment]
> Move to v1.4.0, we can add a document for v1.3.0: #5333 @torchiaf please help survey the reason the common uplink can't be selected from GUI's perspective.

I setup the same environment as above
- harvester-node-0 : default node
- harvester-node-1 : witness node

The API `https://localhost:8006/v1/harvester/network.harvesterhci.io.linkmonitors` response doesn't  have any `havester-node-1` network interface data. 
![api](https://github.com/harvester/harvester/assets/5744158/edda25c1-b7fc-4e09-99c8-5a4b3ef6caf6)
 
 So in frontend codes [updateMatchingNICs()](https://github.com/harvester/dashboard/blob/5a4f2c744210f35892f4e7a24c2fd2565c109de4/pkg/harvester/edit/network.harvesterhci.io.vlanconfig/index.vue#L270-L291) and [intersection()](https://github.com/harvester/dashboard/blob/5a4f2c744210f35892f4e7a24c2fd2565c109de4/pkg/harvester/edit/network.harvesterhci.io.vlanconfig/index.vue#L293-L298)  ,  it doesn't get any NIC that appears both in `harvester-node-0` and `harvester-node-1`  
![log](https://github.com/harvester/harvester/assets/5744158/b813f7dc-34c4-4859-b644-9b4dd3d1cb25)

**Comment by a110605** [comment]
That causes everytime we choose in nodeSelector which contains witness node (harvester-node-1)
- `select all nodes` 
- `select specific node` -> choose `harvester-node-1`

there isn't any nicOptions can be chosen.

![select_all](https://github.com/harvester/harvester/assets/5744158/297b26db-5af6-47aa-9e6d-5e86229212d0)
![select_node1](https://github.com/harvester/harvester/assets/5744158/f7c27e26-7133-4872-8854-fd1e915b5960)

**Comment by lanfon72** [comment]
I can't select the NIC of witness node in version `master-25937fe-head`.

https://github.com/harvester/harvester/assets/5169694/d4b07f63-5002-4a6f-b512-13cc6f945f0f

**Comment by albinsun** [comment]
> I can't select the NIC of witness node in version `master-25937fe-head`.
> 
>  vokoscreenNG-2024-07-03_02-30-43.mp4

Did you create `is_witness` label in advance as it I think it's not a default label?
And filtering the witness node from NIC selection is by-design.
* Ref. harvester/harvester/issues/5743 

cc. @a110605

**Comment by lanfon72** [comment]
It will not match the node labeled `is_witness`, and the additional label `node-role.harvesterhci.io/witness` will not match the node too.

The UI seems broken, other problems like:
1. No labels displayed unless customized label added
![image](https://github.com/harvester/harvester/assets/5169694/24bf7185-4c75-4b32-9934-71f8ad26e474)
![image](https://github.com/harvester/harvester/assets/5169694/3da9dd61-4e0a-4d0d-af5f-c161218591c2)

2. Click `Save` button will not redirect to hosts page (But the change has been applied)
![image](https://github.com/harvester/harvester/assets/5169694/565ebf56-357f-4db8-8a4d-1b6a5cb49fe0)

I will create issues for them.

**Comment by Vicente-Cheng** [comment]
The witness feature needs at least three nodes (two mgmt, one witness).

We only have one mgmt in the two nodes cluster. (prevent the split-brain issue)
Could you try again on the three-node cluster?

**Comment by a110605** [comment]
> I can't select the NIC of witness node in version `master-25937fe-head`.
> 
>  vokoscreenNG-2024-07-03_02-30-43.mp4

[Previous PR](https://github.com/harvester/dashboard/pull/1031) remove witness node from first and second node selector which means user can't choose witness node from these two options

But this screen video sceanrio is user specifies the witness node by giving filter label `kubernetes.io/hostname : harverter-node-1`.

As I mentioned in above [comment](https://github.com/harvester/harvester/issues/5325#issuecomment-2110386816), API doesn't give  =any witness node (harvester-node-1) nic info, that causes when calculating the [matchINCs](https://github.com/harvester/dashboard/blob/a05919686c619b7b65e4b2c1295dd67265cbb20a/pkg/harvester/edit/network.harvesterhci.io.vlanconfig/index.vue#L280-L282) got empty array.

![log1](https://github.com/harvester/harvester/assets/5744158/6f0b0271-86d6-4159-b1de-dc7d4e89e0a9)

In short, I think we can either 
- write down this scenario in doc
- not allow user choose witness node by using third node selector (but not make sense to me)

**Comment by lanfon72** [comment]
okay, It looks I misunderstanding the bug description, 
> the witness node is not able to create network config

it is expected behavior, and this bug describing that NICs would be empty when creating network config with `select all nodes`.

**Comment by a110605** [comment]
> okay, It looks I misunderstanding the bug description,
> 
> > the witness node is not able to create network config
> 
> it is expected behavior, and this bug describing that NICs would be empty when creating network config with `select all nodes`.

By looking at @lanfon72  screen video, I found another bug. 

Try to fix in https://github.com/harvester/dashboard/pull/1062

**Comment by a110605** [comment]
> I can't select the NIC of witness node in version `master-25937fe-head`.
> 
>  vokoscreenNG-2024-07-03_02-30-43.mp4


By looking at @lanfon72  screen video, I found another bug. 

Try to fix in https://github.com/harvester/dashboard/pull/1062  

cc @torchiaf , @Vicente-Cheng

**Comment by a110605** [comment]
Hi @lanfon72, I've fixed the the third node selector (matching rule) logic to avoid user select witness node, could you help verify this ticket again?
See video in https://github.com/harvester/dashboard/pull/1063

**Comment by lanfon72** [comment]
Verified this bug is not fixed. (Case 1)

https://github.com/user-attachments/assets/009514a0-4187-476a-b289-de0305329e30

Test Information
----
* Environment: **qemu/KVM 3 nodes (1 witness)**
* Harvester Version: **v1.4-5fdade5-head**
* **ui-source** Option: **Auto**

### Verify Steps
**prerequisite**
- Install Harvester with at least 1 witness node

**Case 1**
- Follow steps in https://github.com/harvester/harvester/issues/5325#issue-2175586635

**Case 2** (UI only)
- Follow steps in https://github.com/harvester/dashboard/pull/1062

**Comment by a110605** [comment]
> Verified this bug is not fixed. (Case 1)
> 
>  vokoscreenNG-2024-07-17_08-35-30.mp4 
> ## Test Information
> * Environment: **qemu/KVM 3 nodes (1 witness)**
> * Harvester Version: **v1.4-5fdade5-head**
> * **ui-source** Option: **Auto**
> 
> ### Verify Steps
> **prerequisite**
> 
> * Install Harvester with at least 1 witness node
> 
> **Case 1**
> 
> * Follow steps in [[BUG] Unable to select NIC to create network config when cluster contains witness node #5325 (comment)](https://github.com/harvester/harvester/issues/5325#issue-2175586635)
> 
> **Case 2** (UI only)
> 
> * Follow steps in [Fix non-matched rule show matched nodes in create cluster network page dashboard#1062](https://github.com/harvester/dashboard/pull/1062)

Hi @lanfon72 , from your video, @albinsun already mentioned the witness node is added in selected hosts in this [comment](https://github.com/harvester/harvester/issues/5743#issuecomment-2139200834).  We think it's fine since we only filter out  witness node in node selectors UI.

But if we really want to align yaml and UI to only show 2 nodes matched (**harvester-node-0** and **harvester-node-2**).
 I can filter out witness node from selected hosts [here](https://github.com/harvester/dashboard/blob/a0ef90a5a050190e7a9355df4a67a1cdd194120e/pkg/harvester/detail/network.harvesterhci.io.vlanconfig.vue#L63). But for annotation in yaml, it's the response from `/v1/harvester/network.harvesterhci.io.vlanconfigs`, we need backend help to filter out witness node. cc @Vicente-Cheng 


![Screenshot_20240718_211731](https://github.com/user-attachments/assets/a062cc3c-4853-4a9a-a9a8-2b6eddb5785f)

**Comment by a110605** [comment]
hi @lanfon72 , we have merged all the PRs to filter out witness node in network config detail page and YAML . 

Could you help verify again ?

**Comment by lanfon72** [comment]
@Vicente-Cheng, I tested both `v1.3.1-rc1` and `v1.4-146d8ba-head` are not include the change, It looks we didn't bump new version of `network-controller`?

**Comment by Vicente-Cheng** [comment]
> @Vicente-Cheng, I tested both `v1.3.1-rc1` and `v1.4-146d8ba-head` are not include the change, It looks we didn't bump new version of `network-controller`?

Hi @lanfon72, yes you are correct.
We missed the new network-controller.

Please check the v1.3.2 backport issue for the current status (https://github.com/harvester/harvester/issues/5743)
I will move it to ready for testing after the network-controller is updated.

**Comment by lanfon72** [comment]
`v1.4` _network-controller_ not bumped yet (tested on `v1.4-acc153c-head`), move back to **implement**.

**Comment by Vicente-Cheng** [comment]
Sorry @lanfon72,
I did not check the chart version of v1.4-head.

Once the network-controller chart is updated, I will move the status to `Ready for Testing` again.
Thanks.

**Comment by Vicente-Cheng** [comment]
Hi @lanfon72,

It's ready on v1.4-head.
REF: https://github.com/harvester/harvester/blob/v1.4/deploy/charts/harvester/Chart.yaml#L29-L30

The related change on the harvester-network-controller is v0.5.0
Now, the v1.4-head uses the v0.5.1.

So, it should be fine to check this issue again. Thanks!

**Comment by lanfon72** [comment]
Verified this bug has been fixed.


Test Information
----
* Environment: **qemu/KVM 3 nodes**
* Harvester Version: **v1.4-f5fc452-head**
* **ui-source** Option: **Auto**

### Verify Steps
- Same as https://github.com/harvester/harvester/issues/5325#issuecomment-2232071080
